lib/dune_file.ml

```diff
@@ -67,6 +67,143 @@ module Lang = struct
   let duniverse_minimum_version = (1, 11)
 end
 
+module Packages = struct
+  open Sexplib0.Sexp
+  module Set = Set.Make (String)
+  module Map = Map.Make (String)
+
+  type t = Random.State.t
+
+  let init = Random.State.make_self_init
+
+  let random_valid_identifier t =
+    let int_range ~start ~end_ = Random.State.int t (end_ - start) + start in
+    let lower_case_letter_range () = int_range ~start:97 ~end_:122 in
+    (* Seq.init exists only in 4.14+ *)
+    List.init ~len:8 ~f:(fun _ -> Char.chr (lower_case_letter_range ()))
+    |> List.to_seq |> String.of_seq
+
+  let random_public_name v original =
+    let suffix = random_valid_identifier v in
+    (* needs to start with `old.` to be part of the same package *)
+    Fmt.str "%s.%s" original suffix
+
+  let random_library_name v original =
+    let suffix = random_valid_identifier v in
+    (* needs not to have a dot *)
+    Fmt.str "%s_%s" original suffix
+
+  let find_by_name name stanzas =
+    let matches =
+      List.filter_map stanzas ~f:(function
+        | List [ Atom candidate; Atom value ] -> (
+            match String.equal candidate name with
+            | true -> Some value
+            | false -> None)
+        | _ -> None)
+    in
+    match matches with [] -> None | [ x ] -> Some x | _ -> None
+
+  (* type rename_result = { *)
+  (*   changed: bool; *)
+  (*   sexp: Sexplib0.Sexp.t; *)
+  (*   renames: string Map.t; *)
+  (* } *)
+
+  (* determine whether the package should be kept or not, handles [pkg.name] and [pkg] *)
+  let should_keep ~keep name =
+    let test_against =
+      match Base.String.lsplit2 name ~on:'.' with
+      | Some (package, _name) -> package
+      | None -> name
+    in
+    Set.mem test_against keep
+
+  let rename_library t ~keep renames stanzas =
```

Comment:
One trick I recently learned is that you can patch the `dune` file of a subdirectory without writing in that directory: you can mark that directory as `data_only_dirs` and use `(subdir)` to set the contents of the (virtual) dune file in there. I don't know if that applies to that case, but that can be useful if you don't want to change the contents of the unpacked tarball.